### PR TITLE
add `imageplot` (support sixel)

### PR DESCRIPTION
**Pull request by t-bltg**
Fix https://github.com/JuliaPlots/UnicodePlots.jl/issues/153, goes with https://github.com/JuliaPlots/Plots.jl/pull/4011.

**with `Sixel` [^1]**
[^1]: for sixel support, code snippet must be ran under a [compatible terminal](https://github.com/johnnychen94/Sixel.jl#terminals-that-support-sixel).
```julia
using Sixel, UnicodePlots, ImageInTerminal, TestImages, ImageTransformations

main() = begin
  img = imresize(testimage("mandril_color"), ratio=.4)
  display(UnicodePlots.image(img, title="mandril sixel"))
  return
end

main()
```
mandril

**without `Sixel`**
```julia
using UnicodePlots, ImageInTerminal, TestImages, ImageTransformations

main() = begin
  img = imresize(testimage("mandril_color"), ratio=.4)
  ImageInTerminal.use_24bit()
  display(UnicodePlots.image(img, title="mandril 24 bit"))
  ImageInTerminal.use_256()
  display(UnicodePlots.image(img, title="mandril 256 color"))
  return
end

main()
```
mandril_24_256

Discussion:
- ~~naming ?, should we export this ?~~ chose to unexport `image`
-  ~~img is downsized using `restrict` in `ImageInTerminal`, is there a solution to get the ratio so that we can add a centered title and/or labels ?~~ solved using [terminal control sequences](https://invisible-island.net/xterm/ctlseqs/ctlseqs.html) to query the cursor carret size
- is the approach using `@require` standard, for optional dependency `ImageInTerminal`, or should we instead add `ImageInTerminal` as a hard dependency ?
- the sixel implementation was tricky, had to use some ansi escape codes for printing the right border, maybe we can do something better (see the following mwe for writing after a sixel sequence)

```bash
$ printf '\ePq
#0;2;0;0;0#1;2;100;100;0#2;2;0;100;0
#1~~@@vv@@~~@@~~$
#2??}}GG}}??}}??-
#1!14@\e\\\e[A\e[2Cworld
'
```
hi

### CODE REVIEW

**Comment by johnnychen94** [comment]
I hope we can pend this PR for a while until we address https://github.com/JuliaImages/ImageInTerminal.jl/issues/48. The current ImageInTerminal has one unwanted side effect; that any package loads ImageInTerminal would change how colorant arrays are rendered because [it overrides `Base.show`](https://github.com/JuliaImages/ImageInTerminal.jl/blob/4cd5bebf718196cb49a66b600b27d4023696ead4/src/ImageInTerminal.jl#L63-L73).

**Comment by t-bltg** [comment]
> I hope we can pend this PR for a while until we address [JuliaImages/ImageInTerminal.jl#48](https://github.com/JuliaImages/ImageInTerminal.jl/issues/48).

Thanks, for pointing that. Maybe I can help on that matter ?

The current ImageInTerminal has one unwanted side effect; that any package loads ImageInTerminal would change how colorant arrays are rendered in terminal even if you don't call it with UnicodePlots.image because it overrides Base.show.

Is this something we want to change for `ImageInTerminal` 1.0 ?

**Comment by johnnychen94** [comment]
> Is this something we want to change for ImageInTerminal 1.0 ?

Yes, I'd like to separate ImageInTerminal into two packages: backends and frontends. The backend might be as simple as `ascii_encode(img)` and the frontend would provide a Display type and override `Base.display` and `Base.show` method. For instance, Sixel.jl can be seen as an ImageInTerminal backend.

To do this, we need to rework and deprecate the current `imshow` implementation: 1) introduces a pure encoding backend `ascii_encode([io], img, method; kwargs...)` as I did in Sixel.jl, make it a standalone package (say `AsciiImageEncoding`, a bad package name, but you can get the point), and 2) introduce a `TerminalGraphicDisplay` type and keep the previous `Base.show` methods in ImageInTerminal.

By doing this, this PR can directly use the `ascii_encode` methods in the new backend package just like using `sixel_encode` in Sixel. We could also move some core encoding functionality of the UnicodePlots to the new backend package.

**Comment by johnnychen94** [comment]
> Thanks, for pointing that. Maybe I can help on that matter ?

Yeah, given that I have too many items on my todo-list it would be great if you can help implement this! I just sent you an invitation to ImageInTerminal, you can choose to implement the backend package in it as a sub-package like [SnoopCompileCore](https://github.com/timholy/SnoopCompile.jl), or you can create a new package and refactor the encoding part of  ImageInTerminal to the new package.

**Comment by t-bltg** [comment]
Thanks, I'll continue the discussion in https://github.com/JuliaImages/ImageInTerminal.jl/issues/52 about details.

File: Project.toml [hunk 1]
```diff
@@ -25,10 +25,10 @@ StatsBase = "2913bbd2-ae8a-5f71-8c99-4fb6c76f3a91"
 Unitful = "1986cc42-f94f-5a68-af5c-568840ba703d"
 
 [compat]
-ColorSchemes = "3.19"
+ColorSchemes = "≥ 3.19"
```
Comment (johnnychen94) [inline Project.toml:28]: It's not a good idea to allow arbitrary breaking releases with `>=`. You can't ensure this package works with ColorSchemes v4, or Crayons v5 unless they're released.

A lot of Julia packages used this, and then the entire ecosystem shifts to the more restricted version. https://github.com/JuliaRegistries/General/pull/11114

File: Project.toml [hunk 2]
```diff
@@ -18,34 +18,38 @@ LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
...
 Printf = "de0858da-6303-5e67-8744-51eddeeeb8d7"
+Requires = "ae029012-a4dd-5104-9daa-d747884805df"
 SparseArrays = "2f01184e-e22b-5df5-ae63-d93ebab69eaf"
 StaticArrays = "90137ffa-7385-5640-81b9-e52037218182"
 StatsBase = "2913bbd2-ae8a-5f71-8c99-4fb6c76f3a91"
 Unitful = "1986cc42-f94f-5a68-af5c-568840ba703d"
 
 [compat]
-ColorSchemes = "3.19"
+ColorSchemes = "≥ 3.19"
```
Comment (johnnychen94) [inline Project.toml:28]: It's not a good idea to allow arbitrary breaking releases with `>=`. You can't ensure this package works with ColorSchemes v4, or Crayons v5 unless they're released.

A lot of Julia packages used this, and then the entire ecosystem shifts to the more restricted version. https://github.com/JuliaRegistries/General/pull/11114

File: test/tst_imageplot.jl [hunk 3]
```diff
@@ -0,0 +1,21 @@
+if isinteractive()  # FIXME: Sixel.query_terminal(...) fails when !(stdin isa TTY)
+    img = testimage("monarch_color_256")
+
+    @testset "sixel" begin
+        old_enc = ImageInTerminal.ENCODER_BACKEND[]
+        ImageInTerminal.ENCODER_BACKEND[] = :Sixel
+        p = imageplot(img, title = "sixel")
+        @test p.graphics.sixel[]
+        test_ref("imageplot/img_sixel.txt", @show_col(p, :displaysize => T_SZ))
```
Comment (johnnychen94) [inline test/tst_imageplot.jl:9]: It seems that this file is not commited.

Comment (t-bltg) [reply test/tst_imageplot.jl:9]: Xref https://github.com/JuliaIO/Sixel.jl/pull/18#issuecomment-1231525862.

File: Project.toml (hunk 2 above)
Comment (t-bltg) [reply Project.toml:28]: Do you say that all the dependencies like `StaticArrays`, `Unitful`, ... currently at `1` should be pinned to a minor range instead ?

Comment (johnnychen94) [reply Project.toml:28]: > https://pkgdocs.julialang.org/v1/compatibility/#Version-specifier-format
> Caret specifiers are the default and hence 1.2.3 == ^1.2.3. 

I was speaking specifically about the `>=` usage (can't mind reading and realize that `>=` is a typo). Caret is okay, but `^3.19` equivalent to `3.19`.

File: test/tst_imageplot.jl (hunk 3 above)
Comment (johnnychen94) [reply test/tst_imageplot.jl:9]: But if this file is not commited, how would you expect other people to run the test in interactive mode?

File: Project.toml (hunk 2 above)
Comment (t-bltg) [reply Project.toml:28]: I will revert to caret, sorry.  `^3.19` means `[3.19.0, 4.0.0)` according to https://pkgdocs.julialang.org/v1/compatibility/#Caret-specifiers, which is what I originally meant.

File: test/tst_imageplot.jl (hunk 3 above)
Comment (t-bltg) [reply test/tst_imageplot.jl:9]: Yeah, this is a wip since I get different results when running tests in `mlterm ` or `xterm`.

Comment (t-bltg) [reply test/tst_imageplot.jl:9]: I don't know what to do here, should I remove the tests, since non-interactive mod is bound to fail ?

**Comment by t-bltg** [comment]
I will probably merge this in the next few days, unless objection(s).

**Pull request merged**